core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -2178,4 +2181,105 @@ public void testRollGenerationBetweenPrepareCommitAndCommit() throws IOException
+                            translog.currentFileGeneration(),
+                            g -> new ArrayList<>());
+            seqNoForGeneration.add(seqNo);
+            if (rarely()) {
+                translog.rollGeneration();
+            }
+        }
+
+        final Set<Long> seenSeqNos = new HashSet<>();
+        for (final Map.Entry<Long, List<Long>> entry : generations.entrySet()) {
+            final long min = entry.getValue().stream().min(Long::compareTo).orElse(Long.MIN_VALUE);
+            final long max = entry.getValue().stream().max(Long::compareTo).orElse(Long.MAX_VALUE);
+            for (long seqNo = min; seqNo <= max; seqNo++) {
+                if (seenSeqNos.add(seqNo)) {
+                    assertThat(
+                            translog.getMinGenerationForSeqNo(seqNo).translogFileGeneration,
+                            equalTo(entry.getKey()));
+                }
+            }
+        }
+
+        assertThat(seenSeqNos, equalTo(new HashSet<>(seqNos)));
+    }
+
+    public void testSimpleCommit() throws IOException {
+        final int operations = randomIntBetween(1, 4096);
+        long seqNo = 0;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test'"));
+            if (rarely()) {
+                translog.rollGeneration();
+            }
+        }
+
+        final long generation =
+                randomIntBetween(1, Math.toIntExact(translog.currentFileGeneration()));
+        translog.commit(generation);
+        for (long i = 0; i < generation; i++) {
+            assertFileDeleted(translog, i);
+        }
+        for (long i = generation; i <= translog.currentFileGeneration(); i++) {
+            assertFileIsPresent(translog, i);
+        }
+    }
+
+    public void testPrepareCommitAndCommit() throws IOException {
+        final int operations = randomIntBetween(1, 4096);
+        long seqNo = 0;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test"));
+            if (rarely()) {
+                final long generation = translog.currentFileGeneration();
+                translog.prepareCommit();
+                translog.commit(randomIntBetween(1, Math.toIntExact(generation)));
+                for (long g = 0; i < generation; g++) {
+                    assertFileDeleted(translog, g);
+                }
+                for (long g = generation; g < translog.currentFileGeneration(); g++) {
+                    assertFileIsPresent(translog, g);
+                }
+            }
+        }
+    }
+
+    public void testCommitWithOpenView() throws IOException {
+        final int operations = randomIntBetween(1, 1 << 16);
+        long seqNo = 0;
+        long last;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test"));
+            if (rarely()) {
+                try (Translog.View ignored = translog.newView()) {
```

Comment:
I think it will be stronger to hold a view that prevents clearing files even if committed. Right now the view doesn't much does it - we would clean up to `committedGeneration` anyway. I should have written a test like that before - right now we only have the evil testConcurrentWriteViewsAndSnapshot, which we may want to beef up with some rolling and prepare commits as well.